finality-aleph/src/data_io.rs

```diff
@@ -239,11 +239,31 @@ pub(crate) struct DataIO<B: BlockT> {
     pub(crate) metrics: Option<Metrics<B::Header>>,
 }
 
-pub(crate) async fn refresh_best_chain<B: BlockT, SC: SelectChain<B>>(
+pub(crate) async fn refresh_best_chain<B, BE, SC, C>(
     select_chain: SC,
-    best_chain: Arc<Mutex<AlephDataFor<B>>>,
+    client: Arc<C>,
+    data: Arc<Mutex<AlephDataFor<B>>>,
+    max_block_num: NumberFor<B>,
     mut exit: oneshot::Receiver<()>,
-) {
+) where
+    B: BlockT,
+    C: crate::ClientForAleph<B, BE> + Send + Sync + 'static,
+    BE: Backend<B> + 'static,
+    SC: SelectChain<B> + 'static,
+{
+    // A function to reduce block to the level given by num, by traversing down via parents.
+    let reduce_to_num = |header: B::Header, num: NumberFor<B>| {
```

Comment:
I'm pretty sure you can just make this a normal function taking the client as an argument, right?

Reply:
Although see below about the function being needed at all.

Reply:
I could make it into a normal function, but it annoys me that I need to make it generic over 3 arguments (B, C, BE + their bounds). Somehow I thought that this way it's a little bit more elegant, shall I change it?

Reply:
I have a strong preference for code-level functions, which is way stronger than the dislike of multiple template arguments for me. I think this keeps the code much more readable, but it's also just a preference. I would suggest changing it, but it's not a requirement for approval from me.

Reply:
Ok, makes sense to me. Changed to a regular function.